c++/src/io/OutputStream.cc

```diff
@@ -95,9 +91,30 @@ namespace orc {
 
   uint64_t BufferedOutputStream::flush() {
     uint64_t dataSize = dataBuffer->size();
+    // flush data buffer into outputStream
+    if (dataBuffer->getBlockNumber() > 0)
     {
-      SCOPED_STOPWATCH(metrics, IOBlockingLatencyUs, IOCount);
-      outputStream->write(dataBuffer->data(), dataSize);
+      uint64_t ioCount = 0;
+      SCOPED_STOPWATCH(metrics, IOBlockingLatencyUs, nullptr);
+      // try to merge adjacent IO requests
+      BlockBuffer::Block mergeBlock = dataBuffer->getBlock(0);
+      for (uint64_t i = 1; i < dataBuffer->getBlockNumber(); ++i) {
+        auto curBlock = dataBuffer->getBlock(i);
+        if (mergeBlock.data + mergeBlock.size == curBlock.data) {
+          mergeBlock.size += curBlock.size;
+        } else {
+          outputStream->write(mergeBlock.data, mergeBlock.size);
```

Comment:
We may also limit each write not to exceed outputStream->getNaturalWriteSize()

Reply:
The writeTo() of BlockBuffer will limit each write not to exceed NaturalWriteSize of outputStream.